.github/workflows/ci-format.yml

```diff
@@ -36,4 +36,35 @@ jobs:
             -s n \
             --ignore marl_benchmark,examples,scenarios,docs,manager_pb2_grpc.py,worker_pb2_grpc.py \
             --msg-template='{path}: line {line}: {msg_id}: {msg}' \
-            ./smarts ./envision
\ No newline at end of file
+            ./smarts ./envision
+
+  test-types:
+    runs-on: ubuntu-18.04
+    if: github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository
+    container: huaweinoah/smarts:v0.4.18-minimal
+    steps:
+      - name: Install packages
+        run: |
+          add-apt-repository -y ppa:git-core/ppa
+          apt-get update
+          apt-get install -y git
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Install SMARTS
+        run: |
+          python3.7 -m venv ${{env.venv_dir}}
+          . ${{env.venv_dir}}/bin/activate
+          pip install --upgrade pip wheel
+          pip install .[camera-obs,dev,extras,ros,test,train]
+      - name: Get changed files
+        id: changed-files
+        uses: tj-actions/changed-files@v13.2  
+        with:
+          base_sha: origin/develop     
+      - name: Check types
+        if: contains(steps.changed-files.outputs.all_changed_files, '.py')
+        run: |
+          . ${{env.venv_dir}}/bin/activate
+          pytype --disable=pyi-error ${{steps.changed-files.outputs.all_changed_files}}
```

Comment:
So does this mean that if I make a change to, say, `smarts.py`, then `pytype` will be run on that file and I will be forced to fix any existing or lingering type issues there (even if they don't relate to the thing I just changed)?

Or is it only run on _new_ files that I'm adding to the repo?

If it's the former, I'm not sure I would support this.

Reply:
Currently all added, copied, modified, and renamed files, with respect to the `develop` branch will be checked by pytype.

Poor unchecked coding and typing is precisely how we ended up at https://github.com/huawei-noah/SMARTS/pull/1248#discussion_r791071308 and at https://github.com/huawei-noah/SMARTS/pull/1248#discussion_r791097068, which continually bugs the users in a never ending cycle.

pytype is very lenient in type checking. Please give it a try first.

Reply:
I like `pytype` and have been using it now for a few weeks (including fixing old code in unrelated PRs, such as #1219!)  No need to just "give it a try" -- I am using it regularly, including using it to fix old code (see #1208 for example).

My objection is not to pytype itself, but to forcing people making changes/fixes to existing code to have to _also_ change other, completely unrelated code elsewhere in the file that potentially has nothing to do with the Issue on which they're working (and which they may not understand adequately enough to make changes at that point, especially if they are not part of the core smarts team) just to meet the demands of our now-automated checking.  

Note that I'm _not_ saying such fixes shouldn't ever be made:  rather I think they should be made in their own, dedicated PRs, rather than "along the way" in other PRs.  For example, I would argue that even though all of the type hint changes I made in, say, PR #1219 were needed (and I'm glad I did them), they were just a distraction for the actual fixes in that particular PR and made it much harder to review.

Regarding the comments you linked about the `Collisions` class, while I agree with moving it into Events (and said so there), the issue there had _very_ _little_ to do with type hints (which were added retroactively well after both classes were written), and someone might even debate with you that there was anything whatsoever wrong with the way it was.  (I'm not doing that though.)

Where something like `Collision` is defined should _not_ be _solely_ a matter of what uses it within SMARTS.   You may consider it "poor coding" that it was originally defined or remained in `sensors.py` -- I'm not really arguing that one way or another. (I think the circumstances of it remaining  there while `Events` was moved to its own file predate both of us on this project... it's hard for either of us to evaluate that decision without context.)

While I do believe that it _currently_ makes sense to define the `Collision` class in `events.py` in this case, I also believe that it should not be _required_ to be just so by _pytype_ -- that's a decision that has to do with more than _just_ the dependency tree!  

And indeed, I believe the fact that you felt the need to move it -- a breaking change -- in a PR about something entirely different _is a case in point_ about the _misuse_ of the `pytype` tool and/or a misunderstanding of the PR process.

Aside:  In my opinion, this dispute is partly caused by forgetting that type hints are just that -- _hints_.  Because of the fact that the _python import model isn't currently designed for strong typing_, using type hints religiously brings about many spurious issues with circular dependencies that are not real issues, but rather artifacts of python's inability to support forward declarations like C++ does.  In my opinion,  trying to bend over backwards (i.e., make the code more convoluted/complex) in order to accommodate an (imperfect) third-party tool detracts from the value of type hints.  I'd rather leave them off completely in such cases (while using them where they don't interfere).

Reply:
The above is my only outstanding issue with this PR.  I suspect we aren't going to quickly come to an agreement on this point.  So in the interest of getting this PR merged, I suggest that we move the `pytype` CI change to a separate PR on its own (if such a PR doesn't already exist) for further discussion on these points and out of this PR.  If so, I can approve this.

Reply:
> ... I suggest that we move the `pytype` CI change to a separate PR on its own (if such a PR doesn't already exist) ...

Oh, I see I missed reviewing the `pytype` CI PR and that this change has _already_ made it into the `develop` branch.  In that case, I can approve this PR now "as is".

I will bring up my objections to the `pytype` CI change in a meeting and/or Issue instead.

Reply:
Note:  What just happened in PR #1266 is exactly the type of thing I was worried about above.  It was a fairly simple PR about not setting `dones` during reset, but because of the pytype CI reports, Allan felt the need to _also_ attempt to fix the (currently broken) vehicle typing scheme as part of that PR, even though he didn't fully understand the reasons for its being broken or the complexity of the change needed.